## Predicting Outcome of an ODI in Cricket Match using Pandas and SciKit Learn

In [1]:
import platform
import pandas
import sklearn
import os
import numpy as np
import pandas as pd

In [3]:
results = pd.read_csv("C:\Users\Anshul\Desktop\cricket.csv")

In [4]:
results.ix[:5]

teamA        teamB            city home_ground_of  rankTeamA  \
0     Australia  New Zealand      Chandigarh           none          1   
1  South Africa  West Indies          Jaipur           none          4   
2     Australia  West Indies          Mumbai           none          1   
3  South Africa        India          Durban   South Africa          4   
4  South Africa        India       Cape Town   South Africa          4   
5  South Africa        India  Port Elizabeth   South Africa          4   

   pointsTeamA  RatingTeamA  rankTeamB  pointsTeamB  RatingTeamB  \
0         4889          129          3         4284          116   
1         5131          112          8         2471           88   
2         4889          129          8         2471           88   
3         5131          112          2         5278          117   
4         5131          112          2         5278          117   
5         5131          112          2         5278          117   

   winnerNumber    winnerName  
0             1     Australia  
1             8   West Indies  
2             1     Australia  
3             4  South Africa  
4             4  South Africa  
5             4  South Africa

## Let's programatically create a column - 'winnerIsAorB_1or2'; If teamA wins = 1 else 2;

- First I need to Add this new column to the dataframe results, even before I programatically put right values to it.
- Run this only once. And never run afterwards as it will reset all column values to 1 again.
- teamA wins, then = 1, teamA loose or teamB wins = 2, No result = 0

In [5]:
#Run this cell only once. It makes the column values all equal to 1. I have programatically set the values later.
results["winnerIsAorB_1or2"] = 1

In [6]:
for i in range(len(results["teamA"].values)):
    if results["winnerName"][i] == results["teamA"][i]:
        results["winnerIsAorB_1or2"][i] = 1
    elif results["winnerName"][i] == results["teamB"][i]:
        results["winnerIsAorB_1or2"][i] = 2
    else:
        results["winnerIsAorB_1or2"][i] = 0   #To handle the case of no result, this is another class.  

C:\Anaconda2\lib\site-packages\ipykernel\__main__.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  app.launch_new_instance()
C:\Anaconda2\lib\site-packages\ipykernel\__main__.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
C:\Anaconda2\lib\site-packages\ipykernel\__main__.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy


## Yes, the column - 'winnerIsAorB_1or2' has been added to Data Frame 'results'
- See Below for evidence

In [26]:
results.ix[500:590]

teamA         teamB           city home_ground_of  rankTeamA  \
500      Pakistan   West Indies         Mirpur           none          9   
501         India     Australia      Ahmedabad          India          2   
502   New Zealand  South Africa         Mirpur           none          3   
503     Sri Lanka       England        Colombo      Sri Lanka          5   
504     Sri Lanka   New Zealand        Colombo      Sri Lanka          5   
505         India      Pakistan     Chandigarh          India          2   
506         India     Sri Lanka         Mumbai          India          2   
507    Bangladesh     Australia         Mirpur     Bangladesh          7   
508    Bangladesh     Australia         Mirpur     Bangladesh          7   
509    Bangladesh     Australia         Mirpur     Bangladesh          7   
510   West Indies      Pakistan       St Lucia    West Indies          8   
511   West Indies      Pakistan       St Lucia    West Indies          8   
512   West Indies      Pakistan       Barbados    West Indies          8   
513   West Indies      Pakistan       Barbados    West Indies          8   
514   West Indies      Pakistan         Guyana    West Indies          8   
515   West Indies         India       Trinidad    West Indies          8   
516   West Indies         India       Trinidad    West Indies          8   
517   West Indies         India        Antigua    West Indies          8   
518   West Indies         India        Antigua    West Indies          8   
519   West Indies         India        Jamaica    West Indies          8   
520       England     Sri Lanka         London        England          6   
521       England     Sri Lanka          Leeds        England          6   
522       England     Sri Lanka         London        England          6   
523       England     Sri Lanka     Nottingham        England          6   
524       England     Sri Lanka     Manchester        England          6   
525     Sri Lanka     Australia            NaN      Sri Lanka          5   
526      Zimbabwe    Bangladesh            NaN       Zimbabwe         11   
527     Sri Lanka     Australia     Hambantota      Sri Lanka          5   
528      Zimbabwe    Bangladesh            NaN       Zimbabwe         11   
529     Sri Lanka     Australia     Hambantota      Sri Lanka          5   
..            ...           ...            ...            ...        ...   
561      Pakistan     Sri Lanka            NaN           none          9   
562      Pakistan     Sri Lanka      Abu Dhabi           none          9   
563         India   West Indies        Cuttack          India          2   
564    Bangladesh      Pakistan         Mirpur     Bangladesh          7   
565         India   West Indies  Visakhapatnam          India          2   
566    Bangladesh      Pakistan         Mirpur     Bangladesh          7   
567         India   West Indies      Ahmedabad          India          2   
568    Bangladesh      Pakistan     Chittagong     Bangladesh          7   
569         India   West Indies         Indore          India          2   
570         India   West Indies        Chennai          India          2   
571  South Africa     Sri Lanka          Paarl   South Africa          4   
572  South Africa     Sri Lanka    East London   South Africa          4   
573  South Africa     Sri Lanka   Bloemfontein   South Africa          4   
574  South Africa     Sri Lanka      Kimberley   South Africa          4   
575  South Africa     Sri Lanka   Johannesburg   South Africa          4   
576   New Zealand      Zimbabwe        Dunedin    New Zealand          3   
577     Australia         India            NaN      Australia          1   
578   New Zealand      Zimbabwe      Whangarei    New Zealand          3   
579         India     Sri Lanka          Perth           none          2   
580   New Zealand      Zimbabwe         Napier    New Zealand          3   
581   Afghanistan      Pakistan            NaN           non

## The Baseline model creation taking only ranks as feature
- We are taking only the ranks of the teams and comparing its accuracy
- Then we find the accuracy of our baseline model -> 0.3383

In [8]:
def getWinnerTeam(item):
    if item[0] > item[1]:
        return 1
    return 2

In [10]:
teamaranks = results['rankTeamA']
teambranks = results['rankTeamB']
combinedranks = list(zip(teamaranks,teambranks))

In [11]:
winnerList = map(getWinnerTeam,combinedranks)

In [48]:
from sklearn.metrics import accuracy_score
print accuracy_score(winnerList, results['winnerIsAorB_1or2'])

0.338308457711


## The baseline could have been - "Home Team Wins" as in most of the research papers. However we chose "Ranking" for the reasons below.
- This is because unlike other sports like football and soccer where everymatch is played on the Home Ground of either of the teams, Cricket is at times played on Neutral Ground. 
- A neutral ground would mean that it is not the HomeGround of either of the teams.
- So going ahead with rankings.

## Let's do hypothesis testing. 
- Earlier we were skipping proving mathematically, which feature is better? Home Win or Ranking
- If time permits, we will come back to prove it.
- The pseudo code is however there.
## Need to perform Hypothesis Testing as well as Prof. Leman had mentioned
- No need to run the below code snippet right now, it is for future reference only

In [34]:
def getHGWinnerTeam(item):
    if item[0] == item[1]:
        return 1
    elif item[0] == 'none' or item[1] == 'no result':
#         print 'yes'
        return 2
    else:
        return 0

In [16]:
# Will do later. For now let's assume that it has been done.
# teamA = results['teamA']
# teamB = results['teamB']
homeGroundTeam = results['home_ground_of']
winner = results['winnerName']
combinedTeams = list(zip(homeGroundTeam, winner))
combinedTeams

[('none', 'Australia'),
 ('none', 'West Indies'),
 ('none', 'Australia'),
 ('South Africa', 'South Africa'),
 ('South Africa', 'South Africa'),
 ('South Africa', 'South Africa'),
 ('Bangladesh', 'Bangladesh'),
 ('South Africa', 'South Africa'),
 ('Zimbabwe', 'India'),
 ('Bangladesh', 'Bangladesh'),
 ('Sri Lanka', 'Sri Lanka'),
 ('Sri Lanka', 'Pakistan'),
 ('Sri Lanka', 'Pakistan'),
 ('Sri Lanka', 'Sri Lanka'),
 ('Zimbabwe', 'Zimbabwe'),
 ('Zimbabwe', 'New Zealand'),
 ('Zimbabwe', 'New Zealand'),
 ('South Africa', 'South Africa'),
 ('South Africa', 'New Zealand'),
 ('South Africa', 'South Africa'),
 ('Ireland', 'Australia'),
 ('England', 'Australia'),
 ('England', 'Australia'),
 ('England', 'England'),
 ('England', 'England'),
 ('England', 'Australia'),
 ('Zimbabwe', 'Pakistan'),
 ('Zimbabwe', 'Zimbabwe'),
 ('Zimbabwe', 'Pakistan'),
 ('Zimbabwe', 'Zimbabwe'),
 ...]

In [35]:
HGwinner = map(getHGWinnerTeam,combinedTeams)
HGwinner

[2,
 2,
 2,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 0,
 1,
 0,
 1,
 0,
 ...]

In [40]:
count=0
countNeutralVenue = 0
for i in range(len(HGwinner)):
    if HGwinner[i] == 1:
        count = count+1
    elif HGwinner[i] == 2:
        countNeutralVenue = countNeutralVenue+1
#         print 'yes'
# count
print 'Match nt counted towards the HGwin = ',countNeutralVenue


Match nt counted towards the HGwin =  265


## So our hypothesis earlier was that HG has significant role is failing.
- It passed earlier as I ignored these 265 matches that could not contribute to HGwin as 223 were played in Neutral venue and 42 resulted in a Draw.
- There fore, earlier HGSignificance > 50% as 424/(1005-265) = 424/740 =57.29>50%
- Now, we can not ignore those draw matches as well, so include them
- Now the HGSignificance < 50% as 424/1005 = 42.18% < 50%

In [ ]:
# from sklearn.metrics import accuracy_score
# print accuracy_score(winnerList, results['winnerIsAorB_1or2'])

## How many matches happened at "Neutral Venue"?

In [42]:
count_of_neutral_venue_games = 0
for i in range(len(results["teamA"].values)):
    if results["home_ground_of"][i] == 'none':
        count_of_neutral_venue_games = count_of_neutral_venue_games + 1

#Print out the matches that took place at Neutral Venue. THis will help in deciding if Home Ground can bea sifnificant feature
print "Out of 1005 matches, ones that took place at Neutral Venue are - ", count_of_neutral_venue_games

Out of 1005 matches, ones that took place at Neutral Venue are -  223


## Let's add another  column to represent which team (teamA or teamB) were playing on HG
- We will name this column as "home_groundOfAorB_1or2"
- If it is home ground of teamA, then = 1, teamB = 2, if Neutral = 0
- Same like before, the next cell is to be executed only once.

In [43]:
# Execute this line only once 
results["home_groundOfAorB_1or2"] = 1

### Yes It has been added, see evidence below

In [44]:
results.ix[50:55]

teamA         teamB      city home_ground_of  rankTeamA  \
50   Bangladesh         India  Trinidad           none          7   
51      Ireland      Pakistan   Jamaica           none         10   
52  West Indies      Zimbabwe   Jamaica    West Indies          8   
53     Scotland  South Africa  St Kitts           none         14   
54   Bangladesh     Sri Lanka  Trinidad           none          7   
55     Pakistan      Zimbabwe   Jamaica           none          9   

    pointsTeamA  RatingTeamA  rankTeamB  pointsTeamB  RatingTeamB  \
50         3094           96          2         5278          117   
51          549           50          9         3721           87   
52         2471           88         11         1250           45   
53          200           25          4         5131          112   
54         3094           96          5         5811          106   
55         3721           87         11         1250           45   

    winnerNumber    winnerName  winnerIsAorB_1or2  home_groundOfAorB_1or2  
50             7    Bangladesh                  1                       1  
51            10       Ireland                  1                       1  
52             8   West Indies                  1                       1  
53             4  South Africa                  2                       1  
54             5     Sri Lanka                  2                       1  
55             9      Pakistan                  1                       1

## Now let's programatically modify it to hold [0, 1 or 2], as described earlier

In [45]:
for i in range(len(results["teamA"].values)):
    if results["home_ground_of"][i] == results["teamA"][i]:
        results["home_groundOfAorB_1or2"][i] = 1
    elif results["home_ground_of"][i] == results["teamB"][i]:
        results["home_groundOfAorB_1or2"][i] = 2
    else:
        results["home_groundOfAorB_1or2"][i] = 0   #To handle the case of Neutral Venue, this is another class. 

C:\Anaconda2\lib\site-packages\ipykernel\__main__.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
C:\Anaconda2\lib\site-packages\ipykernel\__main__.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  app.launch_new_instance()


## Yes, it has been modified as desired, check below for evidence

In [152]:
results.ix[795:800]  #Taking values from a random range.

teamA        teamB      city home_ground_of  rankTeamA  \
795  Afghanistan     Pakistan  Fatullah           none         12   
796        India    Sri Lanka  Fatullah           none          2   
797  West Indies      England   Antigua    West Indies          8   
798   Bangladesh  Afghanistan  Fatullah     Bangladesh          7   
799        India     Pakistan    Mirpur           none          2   
800  West Indies      England   Antigua    West Indies          8   

     pointsTeamA  RatingTeamA  rankTeamB  pointsTeamB  RatingTeamB  \
795          618           41          9         3721           87   
796         5278          117          5         5811          106   
797         2471           88          6         3968           98   
798         3094           96         12          618           41   
799         5278          117          9         3721           87   
800         2471           88          6         3968           98   

     winnerNumber   winnerName  winnerIsAorB_1or2  home_groundOfAorB_1or2  
795             9     Pakistan                  2                       0  
796             5    Sri Lanka                  2                       0  
797             8  West Indies                  1                       1  
798            12  Afghanistan                  2                       1  
799             9     Pakistan                  2                       0  
800             6      England                  2                       1

### To my surprise, there is no entry as 2 for that column, I tested it. So either it is HG of teamA or Neutral Venue.

In [47]:
for i in range(len(results["teamA"].values)):
    if results["home_groundOfAorB_1or2"][i] == 2:
        print "Found"

## Let's check the F1_Score by keeping features just HomeGround and just as Higher Rank and then will include more features
- Also check the cross Validation Score

In [153]:
from sklearn.metrics import f1_score
features = results["home_groundOfAorB_1or2"].tolist()  ##With assumption that team playing at the home ground wins always. But surprisingly that gives a lo score of 0.35. This is F1_Score though, cross_valscore is around 0.3906
y_true = results["winnerIsAorB_1or2"].values

print("F1 Score using just HomeGround as our feature: {:.4f}".format(f1_score(y_true, features, pos_label=None, average='weighted')))
print("F1 Score using just HigherRankIsWinner as our feature: {:.4f}".format(f1_score(y_true, winnerList, pos_label=None, average='weighted')))
print "So we are choosing HigherRankIsWinner to construct our Baseline Model"


F1 Score using just HomeGround as our feature: 0.3520
F1 Score using just HigherRankIsWinner as our feature: 0.3333
So we are choosing HigherRankIsWinner to construct our Baseline Model


# SPLIT it into Test/Training
- Before we start with classification, let's split the data into Train(80%) and test(20%)
- This need to be done carefully as we just don't simply want to take the first 80% matches. Why? Because what if Scotland just started playing and has not been recorded at all in those 80% matches.
- So take each pairs (14c2) and ensure they have 80% of their data in train set and 20% in testing set.
- Consider Aus Vs. India same as India Vs India.

In [236]:
def getDictWithTeamPairs(results):
    
    dictTeamPairs = {}
    teamA = results['teamA']
    teamB = results['teamB']
    combinedTeamAB = list(zip(teamA, teamB))
    # combinedTeamAB

    for teamPair in combinedTeamAB:
        teamPairAsIs = teamPair[0] +"__"+ teamPair[1]   ## Need to consider India__Australia same as Australi__India
        teamPairRev = teamPair[1] +"__"+ teamPair[0]

        if teamPairAsIs not in dictTeamPairs and teamPairRev not in dictTeamPairs:
            dictTeamPairs[teamPairAsIs] = 1
        elif teamPairRev in dictTeamPairs:
            dictTeamPairs[teamPairRev] = dictTeamPairs[teamPairRev]+1
        else:
            dictTeamPairs[teamPairAsIs] = dictTeamPairs[teamPairAsIs] +1

    return dictTeamPairs



## Deep Copy Saved me

In [235]:
#Make two dicts, one for Training, one for Testnig
# dictTeamPairsTrain80 = dictTeamPairs  ## No this won't work, do deepcopy
# dictTeamPairsTest20 = dictTeamPairs   ## No this won't work, do deepcopy
def getTrainTestTeamPairDicts(dictTeamPairs):
    
    import copy
    dictTeamPairsTrain80 = copy.deepcopy(dictTeamPairs)
    dictTeamPairsTest20 = copy.deepcopy(dictTeamPairs)

    ## Now programatically make the counts = 0 in each of these dicts
    for key, vlaue in dictTeamPairs.items():
        value = dictTeamPairs[key]
        dictTeamPairsTrain80[key] = int(round(value * 0.8) )
        dictTeamPairsTest20[key] = int(round(value * 0.2))
        
    return dictTeamPairsTrain80, dictTeamPairsTest20


## Now I know how many of first Ind_Aus Matches to be kept in Training and the rest of them to be kept in training
- Will Create teo dataframes after deepcopy and keep appending the lines to the respective dataframes.

In [234]:
def splitIntoTrainingAndTesting(results):
    dictTeamPairs = getDictWithTeamPairs(results)
    dictTeamPairsTrain80, dictTeamPairsTest20 = getTrainTestTeamPairDicts(dictTeamPairs)
    
    ## Create Exact copy of the DataFrame
    resultsTrain80 = results.copy(deep=True)
    resultsTest20 = results.copy(deep=True)

    ## Now Iterate over the resluts and keep comparing with the dict values, until the trainDict's value for teamA__teamB reaches 0, 
    ## keep putting 'delete' in teamA and teamB in resultsTest20.
    ## The moment it reaches 0 in trainDict, see if it has already reached 0 in testDict. If yes, u will not even encounter that pair.
    ## If trainDict has reached 0, but testDict is non zero then keep putting 'delete' in resultsTrain80.
    # In the end, take both dataframes and delete all rows where teamA says 'delete'.
    # check the count of both to verify

    for i in range(len(results["teamA"].values)):
        teamPairAsIs = results["teamA"][i] + "__" +  results["teamB"][i]
        teamPairRev = results["teamB"][i] + "__" +  results["teamA"][i]

        if teamPairAsIs in dictTeamPairs:
            teamPairToDelete = teamPairAsIs
        else:
            teamPairToDelete = teamPairRev

        if dictTeamPairsTrain80[teamPairToDelete] > 0:
            dictTeamPairsTrain80[teamPairToDelete] = dictTeamPairsTrain80[teamPairToDelete] - 1
            #Put delete in test wale ke teamA me
            resultsTest20["teamA"][i] = "delete"
        elif dictTeamPairsTrain80[teamPairToDelete] == 0 and dictTeamPairsTest20[teamPairToDelete] > 0:
            dictTeamPairsTest20[teamPairToDelete] = dictTeamPairsTest20[teamPairToDelete] -1
            #Put delete in train wale ke team a me
            resultsTrain80["teamA"][i] = "delete"

    print 'loop ended successfully'

    resultsTrainFinal = resultsTrain80[resultsTrain80.teamA != 'delete']
    resultsTestFinal = resultsTest20[resultsTest20.teamA != 'delete']
    print "length of training", len(resultsTrainFinal)
    print "length of testing", len(resultsTestFinal)
    
    return resultsTrainFinal, resultsTestFinal

## Calling the splitting Data Method

In [237]:
## Kept everything in methods, check if this works?
resultsTrainFinal, resultsTestFinal = splitIntoTrainingAndTesting(results)

C:\Anaconda2\lib\site-packages\ipykernel\__main__.py:28: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
C:\Anaconda2\lib\site-packages\ipykernel\__main__.py:32: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy


loop ended successfully
length of training 807
length of testing 198


# Awesome Awesome Awesome - What a split man!!!!
- Now your training dataFrame is resultsTrainFinal
- Testing Dataframe is resultsTestFinal

# Splitting Ends Here 


## This gives us a very low F1 Score, lets determine cross_val_score
- Feature 1 --> Home_ground_Information (Cross Val Score = 0.3906)
- Feature 2 --> Ranking of team A only (Cross Val Score = 0.5370)
- Feature 3 --> Ranking of team B only (Cross Val Score = 0.5659)
- Feature 4 --> Two featutes, rank of Team A and of team B (Cross Val Score = 0.5601)  Reduces after including Team A ranking
- Feature 5 --> Home Ground info and Rank of team A (Score = 0.5352)
- Feature 6 --> Home Ground info and Rank of team B (Score = 0.5529) 
- Feature 7 --> Home Ground info and Rank of team A and of B (Score = 0.5497)

- So Better do not include ranking of team A

# Prediction
## Features = Ranks of both teams + HomeGroundOfWhichTeam
## Classifier = Decision Tree Classifier

In [154]:
from sklearn.tree import DecisionTreeClassifier
clf = DecisionTreeClassifier(random_state=14)

In [378]:
from sklearn.cross_validation import cross_val_score

# features = resultsTrainFinal[["rankTeamB", "rankTeamA"]].values #This is fine
# features = results[["rankTeamA", "rankTeamB"]].values #This is fine, and same as above
# features = results[["home_groundOfAorB_1or2"]].values #This just takes the home ground as the feature
# features = resultsTrainFinal[["rankTeamA", "rankTeamB","home_groundOfAorB_1or2"]].values #Takes all three features. Note I am going to introduce more features. This is a baseline model
y_true = resultsTrainFinal[["winnerIsAorB_1or2"][0]].values
y_true_test = resultsTestFinal[["winnerIsAorB_1or2"][0]].values
# print features type is ndarray, but to list is also fine
# print type(y_true) is nd array
clf = DecisionTreeClassifier(random_state=14)

scores = cross_val_score(clf, features, y_true)
print "Here is the cross val score for above sets of features"
print "F1: {0:.4f}".format(np.mean(scores))

# print y_true

Here is the cross val score for above sets of features
F1: 0.5043


In [212]:
clf.fit(features, y_true)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None, min_samples_leaf=1,
            min_samples_split=2, min_weight_fraction_leaf=0.0,
            random_state=14, splitter='best')

In [213]:
featureVector = resultsTestFinal[["rankTeamA", "rankTeamB","home_groundOfAorB_1or2"]].values
predictedVector = resultsTestFinal[["winnerIsAorB_1or2"][0]].values
# print featureVector[1]
# print predictedVector

correctlyPredicted = 0 
for i in range(len(resultsTestFinal["teamA"].values)):
    input = featureVector[i]
    if clf.predict(input)[0] == predictedVector[i]:
        correctlyPredicted = correctlyPredicted +1
#     print clf.predict(input)[0], predictedVector[i]

accuracy_on_testData =  correctlyPredicted/float(len(resultsTestFinal["teamA"].values))
print "Accuracy on test data using Decision Tree Classifier with ranks of teams and homeground as features = ", accuracy_on_testData 

    

Accuracy on test data using Decision Tree Classifier with ranks of teams and homeground as features =  0.575757575758


# Let's try to predict the test accuracy by predicting the results for this test data

## The predicted results above are biased more towards which team is ranked higher. 
- It almost ignores the HomeGround win feature. And to me it makes sense as well. 
- This is because when we have only two features namely rank of Team A and of team B (Cross Val Score = 0.5601) where as if we add homeground wins as the third feature, it reduces the overall score to (Score = 0.5497)

## Let Add more features, "teamALastWin" and "teamBLastWin" (True/False)

In [215]:
# Execute this only once in order to add this feature, then modify it programatically.
results["teamALastWin"] = True
results["teamBLastWin"] = True

In [216]:
results.ix[795:800]  #Taking values from a random range.

teamA        teamB      city home_ground_of  rankTeamA  \
795  Afghanistan     Pakistan  Fatullah           none         12   
796        India    Sri Lanka  Fatullah           none          2   
797  West Indies      England   Antigua    West Indies          8   
798   Bangladesh  Afghanistan  Fatullah     Bangladesh          7   
799        India     Pakistan    Mirpur           none          2   
800  West Indies      England   Antigua    West Indies          8   

     pointsTeamA  RatingTeamA  rankTeamB  pointsTeamB  RatingTeamB  \
795          618           41          9         3721           87   
796         5278          117          5         5811          106   
797         2471           88          6         3968           98   
798         3094           96         12          618           41   
799         5278          117          9         3721           87   
800         2471           88          6         3968           98   

     winnerNumber   winnerName  winnerIsAorB_1or2  home_groundOfAorB_1or2  \
795             9     Pakistan                  2                       0   
796             5    Sri Lanka                  2                       0   
797             8  West Indies                  1                       1   
798            12  Afghanistan                  2                       1   
799             9     Pakistan                  2                       0   
800             6      England                  2                       1   

    teamALastWin teamBLastWin  
795         True         True  
796         True         True  
797         True         True  
798         True         True  
799         True         True  
800         True         True

In [217]:
#Now let's programatically modify two newly added features A) teamALastWin B) teamBLastWin.
#If they won their last match irrespective of who they played it against, then the value corresponding to that row will be true for that team, else false
dictTeams={}
distinctTeams = results["teamA"].unique()

## Add them to dictionary dictTeams and Initialize everyteam with False-
for i in range(len(distinctTeams)):
    dictTeams[distinctTeams[i]] = False

print dictTeams

{'Pakistan': False, 'Australia': False, 'England': False, 'Bangladesh': False, 'South Africa': False, 'Afghanistan': False, 'Scotland': False, 'India': False, 'Sri Lanka': False, 'Zimbabwe': False, 'Ireland': False, 'United Arab Emirates': False, 'West Indies': False, 'New Zealand': False}


### So far, we have added all teams in dictTeams and initialized their last wins to be False
#### Now Let's modify the column with their actual last win after determining this from the data

In [238]:
for i in range(len(results["teamA"].values)):
    teamA = results["teamA"][i]
    teamB = results["teamB"][i]
    winner = results["winnerName"][i]
    
    results["teamALastWin"][i] = dictTeams[teamA]
    results["teamBLastWin"][i] = dictTeams[teamB]
    
    
    # Update dict with current match to hold the latest wins of teamA and teamB
    if winner == teamA:
        dictTeams[teamA] = True
        dictTeams[teamB] = False
    elif winner == teamB:
        dictTeams[teamA] = False
        dictTeams[teamB] = True
        #else, the match has been drawn with winner = no result, so need not to update the dict in this case.

C:\Anaconda2\lib\site-packages\ipykernel\__main__.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
C:\Anaconda2\lib\site-packages\ipykernel\__main__.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy


## Added a new feature? U will have to split the dataset again to obtain latest Training and Testing Dataset.

In [239]:
resultsTrainFinal, resultsTestFinal = splitIntoTrainingAndTesting(results)

C:\Anaconda2\lib\site-packages\ipykernel\__main__.py:28: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
C:\Anaconda2\lib\site-packages\ipykernel\__main__.py:32: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy


loop ended successfully
length of training 807
length of testing 198


### Yes, it has been modified and it looks perfectly fine, I have verified, check below for evidence

In [251]:
# results.ix[:5]  #Taking values from a random range.
# resultsTrainFinal.ix[:5]
resultsTestFinal.ix[:]


teamA         teamB               city home_ground_of  rankTeamA  \
384    New Zealand     Australia         Wellington    New Zealand          3   
397       Zimbabwe     Sri Lanka                NaN       Zimbabwe         11   
478      Australia   New Zealand             Nagpur           none          1   
479     Bangladesh       Ireland             Mirpur     Bangladesh          7   
489      Sri Lanka      Zimbabwe                NaN      Sri Lanka          5   
490     Bangladesh       England         Chittagong     Bangladesh          7   
509     Bangladesh     Australia             Mirpur     Bangladesh          7   
588        England      Pakistan                NaN           none          6   
590        England      Pakistan                NaN           none          6   
635      Sri Lanka         India            Colombo      Sri Lanka          5   
636      Sri Lanka         India                NaN      Sri Lanka          5   
643        England  South Africa             London        England          6   
645        England  South Africa         Nottingham        England          6   
657          India      Pakistan              Delhi          India          2   
660      Australia     Sri Lanka                NaN      Australia          1   
662      Australia     Sri Lanka           Brisbane      Australia          1   
665      Australia     Sri Lanka                NaN      Australia          1   
667      Australia     Sri Lanka             Hobart      Australia          1   
671      Australia   West Indies              Perth      Australia          1   
672      Australia   West Indies              Perth      Australia          1   
673      Australia   West Indies           Canberra      Australia          1   
674      Australia   West Indies                NaN      Australia          1   
675      Australia   West Indies                NaN      Australia          1   
680    West Indies      Zimbabwe            Grenada    West Indies          8   
690       Zimbabwe    Bangladesh           Bulawayo       Zimbabwe         11   
691       Zimbabwe    Bangladesh           Bulawayo       Zimbabwe         11   
704      Australia   New Zealand         Birmingham           none          1   
707          India      Pakistan         Birmingham           none          2   
709      Australia     Sri Lanka             London           none          1   
710        England  South Africa             London        England          6   
...            ...           ...                ...            ...        ...   
970        England   New Zealand  Chester-le-Street        England          6   
971     Bangladesh         India             Mirpur     Bangladesh          7   
972     Bangladesh         India             Mirpur     Bangladesh          7   
975      Sri Lanka      Pakistan                NaN      Sri Lanka          5   
976     Bangladesh  South Africa             Mirpur     Bangladesh          7   
977       Zimbabwe         India                NaN       Zimbabwe         11   
978       Zimbabwe         India                NaN       Zimbabwe         11   
979     Bangladesh  South Africa         Chittagong     Bangladesh          7   
980      Sri Lanka      Pakistan                NaN      Sri Lanka          5   
981      Sri Lanka      Pakistan            Colombo      Sri Lanka          5   
982      Sri Lanka      Pakistan            Colombo      Sri Lanka          5   
983      Sri Lanka      Pakistan         Hambantota      Sri Lanka          5   
985       Zimbabwe   New Zealand                NaN       Zimbabwe         11   
986       Zimbabwe   New Zealand                NaN       Zimbabwe         11   
987   South Africa   New Zealand          Centurion   South Africa          4   
988   South Africa   New Zealand      Potchefstroom   South Africa          4   
989   South Africa   New Zealand             Durban   South Africa          4   
990        Ireland     Australia            

### Let See if the cross_val_score improved or not? oh snap! It actually reduces after including some specific features
- We will deivce an algorithm to improve the feature selection, and then will come up with best possible moodel on this

## rakesh Jumped in
- Taking out all the numerical features only. Right now the available features doe not include the one hot encoding of teams.
- Ranks of teams, Pints, Ratings, HomGroung info, Team's Last wins against any opponent.

## Future addition of features
- Add who won the match (teamA or teamB) in their most recent encounter in the past. 1 for A, 2 for B and 0 for the draw.


In [371]:
# trainDf = results.ix[:,[4,5,6,7,8,9,13,14,15]]
# trainDf = resultsTrainFinal.ix[:] #,[4,5,6,7,8,9,13,14,15]]
# trainDf = resultsTrainFinal.ix[:,[4,7,13,14,15]]  # Ranks, HomeGrnd, TeamLastWin
# trainDf = resultsTrainFinal.ix[:,[4,5,7,8,13,14,15]]  # Ranks, HomeGrnd, TeamLastWin, PointsGainedByTeams
trainDf = resultsTrainFinal.ix[:,[4,5,6,7,8,9,13,14,15]] # Ranks, HomeGrnd, TeamLastWin, PointsGainedByTeams, RatingOfTeam
testDf = resultsTestFinal.ix[:,[4,5,6,7,8,9,13,14,15]] # Ranks, HomeGrnd, TeamLastWin, PointsGainedByTeams, RatingOfTeam

trainDf

rankTeamA  pointsTeamA  RatingTeamA  rankTeamB  pointsTeamB  \
0             1         4889          129          3         4284   
1             4         5131          112          8         2471   
2             1         4889          129          8         2471   
3             4         5131          112          2         5278   
4             4         5131          112          2         5278   
5             4         5131          112          2         5278   
6             7         3094           96         11         1250   
7             4         5131          112          2         5278   
8             7         3094           96         11         1250   
9             9         3721           87          8         2471   
10            7         3094           96         11         1250   
11            7         3094           96         11         1250   
12            9         3721           87          8         2471   
13            9         3721           87          8         2471   
14            9         3721           87          8         2471   
15            3         4284          116          5         5811   
16            3         4284          116          5         5811   
17            3         4284          116          5         5811   
18            1         4889          129          6         3968   
19            1         4889          129          3         4284   
20            6         3968           98          3         4284   
21            1         4889          129          6         3968   
22            1         4889          129          3         4284   
23            6         3968           98          3         4284   
24            1         4889          129          6         3968   
25            2         5278          117          8         2471   
26            1         4889          129          3         4284   
27            6         3968           98          3         4284   
28            2         5278          117          8         2471   
29            1         4889          129          6         3968   
...         ...          ...          ...        ...          ...   
891           4         5131          112          8         2471   
892          12          618           41         10          549   
893           3         4284          116          5         5811   
895           4         5131          112          8         2471   
915           3         4284          116         14          200   
916          12          618           41          7         3094   
917          13          300           32         11         1250   
920          12          618           41          5         5811   
922           6         3968           98         14          200   
924          10          549           50         13          300   
928           2         5278          117         13          300   
931           9         3721           87         11         1250   
933           1         4889          129         12          618   
934           9         3721           87         13          300   
935           7         3094           96         14          200   
937          10          549           50         11         1250   
940           3         4284          116         12          618   
943          14          200           25          5         5811   
944           4         5131          112         13          300   
945          12          618           41          6         3968   
947           1         4889          129         14          200   
948           2         5278          117         11         1250   
950          13          300           32          8         2471   
962           9         3721           87         11         1250   
963           9         3721           87         11         1250   
973           7         3094           96          4         5131   
974 

In [260]:
from sklearn.cross_validation import cross_val_score

# features = results[["teamALastWin"]].values                                                      #Score --> 0.4245
# features = results[["teamBLastWin"]].values                                                      #Score --> 0.4274
# features = results[["teamALastWin", "teamBLastWin"]].values                                      #Score --> 0.4800
# features = results[["rankTeamA", "teamALastWin", "teamBLastWin"]].values                         #Score --> 0.5432
# features = results[["rankTeamB", "teamALastWin", "teamBLastWin"]].values                         #Score --> 0.5546 
# features = results[["rankTeamA", "rankTeamB",  "teamALastWin", "teamBLastWin"]].values           #Score --> 0.5151 (Reduced when rankTeamA and rankTeamB are feede together)

# features = results[["home_groundOfAorB_1or2", "rankTeamB", "teamALastWin", "teamBLastWin"]].values          # --> 0.5405
# features = results[["home_groundOfAorB_1or2", "rankTeamA", "rankTeamB", "teamALastWin", "teamBLastWin"]].values    # --> 0.4998
# # So the best feature combination so far has been "rankTeamB", "teamALastWin", "teamBLastWin"
# However, rankTeamB alone had given better score, infact the best so far of 0.5659


clf = DecisionTreeClassifier(random_state=14)

# scores = cross_val_score(clf, features, y_true,scoring=scorer) ## This was using the above feature list
scores = cross_val_score(clf, trainDf, y_true,scoring=scorer)  ## This is using the newly created DF from Rakesh.
print "Here is the cross val score for above sets of features"
print "F1: {0:.4f}".format(np.mean(scores))

Here is the cross val score for above sets of features
F1: 0.5003


## Feature Importance by Decision Tree Classifier

In [268]:
clf.fit(trainDf, y_true)
clf.feature_importances_

array([ 0.08693525,  0.22469557,  0.04775404,  0.14454615,  0.15996553,
        0.09024859,  0.07256618,  0.08562546,  0.08766324])

In [269]:
import matplotlib.pyplot as plt
import matplotlib
matplotlib.style.use('ggplot')
feature_importance = pd.Series(clf.feature_importances_, index=trainDf.columns)
feature_importance.sort_values(inplace=True)
print feature_importance, trainDf.columns
print len(trainDf.columns)
feature_importance.plot(kind='barh', figsize=(7,6));
trainDf.plot()
plt.show()

RatingTeamA               0.047754
home_groundOfAorB_1or2    0.072566
teamALastWin              0.085625
rankTeamA                 0.086935
teamBLastWin              0.087663
RatingTeamB               0.090249
rankTeamB                 0.144546
pointsTeamB               0.159966
pointsTeamA               0.224696
dtype: float64 Index([u'rankTeamA', u'pointsTeamA', u'RatingTeamA', u'rankTeamB',
       u'pointsTeamB', u'RatingTeamB', u'home_groundOfAorB_1or2',
       u'teamALastWin', u'teamBLastWin'],
      dtype='object')
9


## Our algo to find the best combination of features giving highest score is ready.
## So let's focus on adding more features
## Let's start with the encoded feature. 

In [261]:
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
encoding = LabelEncoder()
encoding.fit(results["teamA"].values)

LabelEncoder()

In [262]:
team_A_encoded = encoding.transform(results["teamA"].values)
team_B_encoded = encoding.transform(results["teamB"].values)
X_teams = np.vstack([team_A_encoded, team_B_encoded]).T      # .T makes it look tabular
X_teams[:5], X_teams.shape

(array([[ 1,  6],
        [ 9, 12],
        [ 1, 12],
        [ 9,  4],
        [ 9,  4]], dtype=int64), (1005L, 2L))

## One Hot EnCoding

In [263]:
onehot = OneHotEncoder()
X_teams = onehot.fit_transform(X_teams).todense()
X_teams.shape

print("teamA:", X_teams[0,:14])
print("teamB:", X_teams[0,14:])

('teamA:', matrix([[ 0.,  1.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
          0.]]))
('teamB:', matrix([[ 0.,  0.,  0.,  0.,  0.,  0.,  1.,  0.,  0.,  0.,  0.,  0.,  0.,
          0.]]))


In [265]:
# Declared 14x2 columns to hold the one hot encoded teamA and teamB. 14 each since there are 14 disticnt teams.
columns = ['teamA1','teamA2', 'teamA3', 'teamA4', 'teamA5', 'teamA6', 'teamA7', 'teamA8', 'teamA9', 'teamA10', 'teamA11', 'teamA12', 'teamA13', 'teamA14',
           'teamB1','teamB2', 'teamB3', 'teamB4', 'teamB5', 'teamB6', 'teamB7', 'teamB8', 'teamB9', 'teamB10', 'teamB11', 'teamB12', 'teamB13', 'teamB14']

# Create a numpy array of size 1005 x 28
data = np.array([np.arange(1005)]*28).T   # Have hardCoded for simplicity to relate easily for the reader. I know it is not a good practice, I know the workaround.

# Creating the dataframe out of this data using those columns
df = pd.DataFrame(data, columns=columns)

df.ix[:5]


teamA1  teamA2  teamA3  teamA4  teamA5  teamA6  teamA7  teamA8  teamA9  \
0       0       0       0       0       0       0       0       0       0   
1       1       1       1       1       1       1       1       1       1   
2       2       2       2       2       2       2       2       2       2   
3       3       3       3       3       3       3       3       3       3   
4       4       4       4       4       4       4       4       4       4   
5       5       5       5       5       5       5       5       5       5   

   teamA10   ...     teamB5  teamB6  teamB7  teamB8  teamB9  teamB10  teamB11  \
0        0   ...          0       0       0       0       0        0        0   
1        1   ...          1       1       1       1       1        1        1   
2        2   ...          2       2       2       2       2        2        2   
3        3   ...          3       3       3       3       3        3        3   
4        4   ...          4       4       4       4       4        4        4   
5        5   ...          5       5       5       5       5        5        5   

   teamB12  teamB13  teamB14  
0        0        0        0  
1        1        1        1  
2        2        2        2  
3        3        3        3  
4        4        4        4  
5        5        5        5  

[6 rows x 28 columns]

### Let's programatically fill the encoding to these columns

In [266]:
#Get list of all columns in DataFrame df
column_header = list(df.columns.values)


for i in range(1005):
    for column in range(len(column_header)):
        df[column_header[column]][i] = int(X_teams[i].tolist()[0][column])

## Yes, it has been added, see below for evidence.

In [267]:
df.ix[:115]

teamA1  teamA2  teamA3  teamA4  teamA5  teamA6  teamA7  teamA8  teamA9  \
0         0       1       0       0       0       0       0       0       0   
1         0       0       0       0       0       0       0       0       0   
2         0       1       0       0       0       0       0       0       0   
3         0       0       0       0       0       0       0       0       0   
4         0       0       0       0       0       0       0       0       0   
5         0       0       0       0       0       0       0       0       0   
6         0       0       1       0       0       0       0       0       0   
7         0       0       0       0       0       0       0       0       0   
8         0       0       1       0       0       0       0       0       0   
9         0       0       0       0       0       0       0       1       0   
10        0       0       1       0       0       0       0       0       0   
11        0       0       1       0       0       0       0       0       0   
12        0       0       0       0       0       0       0       1       0   
13        0       0       0       0       0       0       0       1       0   
14        0       0       0       0       0       0       0       1       0   
15        0       0       0       0       0       0       1       0       0   
16        0       0       0       0       0       0       1       0       0   
17        0       0       0       0       0       0       1       0       0   
18        0       1       0       0       0       0       0       0       0   
19        0       1       0       0       0       0       0       0       0   
20        0       0       0       1       0       0       0       0       0   
21        0       1       0       0       0       0       0       0       0   
22        0       1       0       0       0       0       0       0       0   
23        0       0       0       1       0       0       0       0       0   
24        0       1       0       0       0       0       0       0       0   
25        0       0       0       0       1       0       0       0       0   
26        0       1       0       0       0       0       0       0       0   
27        0       0       0       1       0       0       0       0       0   
28        0       0       0       0       1       0       0       0       0   
29        0       1       0       0       0       0       0       0       0   
..      ...     ...     ...     ...     ...     ...     ...     ...     ...   
86        0       0       1       0       0       0       0       0       0   
87        0       0       1       0       0       0       0       0       0   
88        0       0       0       0       0       1       0       0       0   
89        0       0       0       0       0       1       0       0       0   
90        0       0       0       0       1       0       0       0       0   
91        0       0       0       0       1       0       0       0       0   
92        0       0       0       1       0       0       0       0       0   
93        0       0       0       0       1       0       0       0       0   
94        0       0       0       1       0       0       0       0       0   
95        0       0       0       1       0       0       0       0       0   
96        0       0       0       0       0       1       0       0       0   
97        0       0       0       0       0       0       0       0       0   
98        0       0       0       0       0       0       0       0       0   
99        0       0       0       0       0       0       0       0       0   
100       0       0       0       0       0       0       0       0       1   
101       0       0       0       1       0       0       0       0       0   
102       0       0       0       0       0       0       0       0       0   
103       0       0       0       1       0       0       0       0       0   
104       0       0       0       0       0       0    

# Now need to add these 28 columns to our original dataframe of results-
## Actually not into results but make a copy of results and leave results untouched.

In [280]:
hotencodedDf = results.copy(deep=True)

for i in range(14):
    hotencodedDf["teamA" + str(i+1)] = df["teamA" + str(i+1)]
    
for i in range(14):
    hotencodedDf["teamB" + str(i+1)] = df["teamB" + str(i+1)]

hotencodedDf.ix[:4,4:]  ## Careful, u included the winner column too

rankTeamA  pointsTeamA  RatingTeamA  rankTeamB  pointsTeamB  RatingTeamB  \
0          1         4889          129          3         4284          116   
1          4         5131          112          8         2471           88   
2          1         4889          129          8         2471           88   
3          4         5131          112          2         5278          117   
4          4         5131          112          2         5278          117   

   winnerNumber    winnerName  winnerIsAorB_1or2  home_groundOfAorB_1or2  \
0             1     Australia                  1                       0   
1             8   West Indies                  2                       0   
2             1     Australia                  1                       0   
3             4  South Africa                  1                       1   
4             4  South Africa                  1                       1   

    ...    teamB5 teamB6  teamB7  teamB8  teamB9  teamB10  teamB11  teamB12  \
0   ...         0      0       1       0       0        0        0        0   
1   ...         0      0       0       0       0        0        0        0   
2   ...         0      0       0       0       0        0        0        0   
3   ...         1      0       0       0       0        0        0        0   
4   ...         1      0       0       0       0        0        0        0   

   teamB13  teamB14  
0        0        0  
1        1        0  
2        1        0  
3        0        0  
4        0        0  

[5 rows x 40 columns]

# You are ready to test but have you splitted the data yet?? So do it now.

In [281]:
train, test = splitIntoTrainingAndTesting(hotencodedDf)
# And make deep copy to avoid any confusion

trainDfAfterEncoding = train.copy(deep = True)
testDfAfterEncoding = test.copy(deep = True)
# hotencodedDf.ix[:4,4:]


C:\Anaconda2\lib\site-packages\ipykernel\__main__.py:28: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
C:\Anaconda2\lib\site-packages\ipykernel\__main__.py:32: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy


loop ended successfully
length of training 807
length of testing 198


In [328]:
trainDfAfterEncoding.ix[:5]

teamA        teamB            city home_ground_of  rankTeamA  \
0     Australia  New Zealand      Chandigarh           none          1   
1  South Africa  West Indies          Jaipur           none          4   
2     Australia  West Indies          Mumbai           none          1   
3  South Africa        India          Durban   South Africa          4   
4  South Africa        India       Cape Town   South Africa          4   
5  South Africa        India  Port Elizabeth   South Africa          4   

   pointsTeamA  RatingTeamA  rankTeamB  pointsTeamB  RatingTeamB   ...     \
0         4889          129          3         4284          116   ...      
1         5131          112          8         2471           88   ...      
2         4889          129          8         2471           88   ...      
3         5131          112          2         5278          117   ...      
4         5131          112          2         5278          117   ...      
5         5131          112          2         5278          117   ...      

   teamB5 teamB6  teamB7  teamB8 teamB9 teamB10  teamB11  teamB12  teamB13  \
0       0      0       1       0      0       0        0        0        0   
1       0      0       0       0      0       0        0        0        1   
2       0      0       0       0      0       0        0        0        1   
3       1      0       0       0      0       0        0        0        0   
4       1      0       0       0      0       0        0        0        0   
5       1      0       0       0      0       0        0        0        0   

   teamB14  
0        0  
1        0  
2        0  
3        0  
4        0  
5        0  

[6 rows x 44 columns]

In [313]:
trainDfAfterEncodingAsFeatures = trainDfAfterEncoding.ix[:,[4,5,6,7,8,9,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43]]
trainDfAfterEncodingAsFeatures.head()

testDfAfterEncodingAsFeatures = testDfAfterEncoding.ix[:,[4,5,6,7,8,9,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43]]
testDfAfterEncodingAsFeatures.head()

features = trainDfAfterEncodingAsFeatures.values
features

array([[1L, 4889L, 129L, ..., 0, 0, 0],
       [4L, 5131L, 112L, ..., 0, 1, 0],
       [1L, 4889L, 129L, ..., 0, 1, 0],
       ..., 
       [11L, 1250L, 45L, ..., 0, 0, 0],
       [11L, 1250L, 45L, ..., 0, 0, 0],
       [11L, 1250L, 45L, ..., 0, 0, 0]], dtype=object)

## Let's try to find the best model with these features by using the combination method from collections in python

In [306]:
import itertools

feature_list = ["home_groundOfAorB_1or2","rankTeamA", "rankTeamB","teamALastWin","teamBLastWin", "pointsTeamA", "pointsTeamB","RatingTeamA","RatingTeamB"]
# encoded_team_list = ['teamA1','teamB1']
encoded_team_list = ['teamA1','teamA2', 'teamA3', 'teamA4', 'teamA5', 'teamA6', 'teamA7', 'teamA8', 'teamA9', 'teamA10', 'teamA11', 'teamA12', 'teamA13', 'teamA14', 'teamB1','teamB2', 'teamB3', 'teamB4', 'teamB5', 'teamB6', 'teamB7', 'teamB8', 'teamB9', 'teamB10', 'teamB11', 'teamB12', 'teamB13', 'teamB14']


lst = feature_list
combs = []
score_list=[]

for i in xrange(1, len(lst)+1):
    els = [list(x) for x in itertools.combinations(lst, i)]
    combs.extend(els)

    
# for this_combination in combs:
#     print this_combination
#     del this_combination[:]

for this_combination in combs:
    print this_combination
    for j in range(len(encoded_team_list)):
        this_combination.append(encoded_team_list[j])
        
    features = trainDfAfterEncodingAsFeatures[this_combination].values
    
    clf = DecisionTreeClassifier(random_state=14)
    scores = cross_val_score(clf, features, y_true)
    print np.mean(scores), len(this_combination), this_combination
#     if np.mean(scores) == 0.57625158341135652:
#         print '#########################################################'           
#         print this_combination
#         break
    combi = this_combination
    score_list.append((np.mean(scores), len(this_combination), this_combination))
    
    del this_combination[:]

for i in range(len(score_list)):
    print score_list[i]

print "\n Now sorted list on scores"

sorted_by_scores = sorted(score_list, key=lambda tup: tup[0])
# del score_list[:]

for i in range(len(sorted_by_scores)):
    print sorted_by_scores[i]


['home_groundOfAorB_1or2']
0.535303188853 29 ['home_groundOfAorB_1or2', 'teamA1', 'teamA2', 'teamA3', 'teamA4', 'teamA5', 'teamA6', 'teamA7', 'teamA8', 'teamA9', 'teamA10', 'teamA11', 'teamA12', 'teamA13', 'teamA14', 'teamB1', 'teamB2', 'teamB3', 'teamB4', 'teamB5', 'teamB6', 'teamB7', 'teamB8', 'teamB9', 'teamB10', 'teamB11', 'teamB12', 'teamB13', 'teamB14']
['rankTeamA']
0.562569991372 29 ['rankTeamA', 'teamA1', 'teamA2', 'teamA3', 'teamA4', 'teamA5', 'teamA6', 'teamA7', 'teamA8', 'teamA9', 'teamA10', 'teamA11', 'teamA12', 'teamA13', 'teamA14', 'teamB1', 'teamB2', 'teamB3', 'teamB4', 'teamB5', 'teamB6', 'teamB7', 'teamB8', 'teamB9', 'teamB10', 'teamB11', 'teamB12', 'teamB13', 'teamB14']
['rankTeamB']
0.56382754126 29 ['rankTeamB', 'teamA1', 'teamA2', 'teamA3', 'teamA4', 'teamA5', 'teamA6', 'teamA7', 'teamA8', 'teamA9', 'teamA10', 'teamA11', 'teamA12', 'teamA13', 'teamA14', 'teamB1', 'teamB2', 'teamB3', 'teamB4', 'teamB5', 'teamB6', 'teamB7', 'teamB8', 'teamB9', 'teamB10', 'teamB11', 

## The above gives best result of 0.57625 with featuresa as -
### 'rankTeamB', 'pointsTeamA', 'RatingTeamB' and one hot encoding of team names.

## Need to fit and predict.
- Fitting the model with rankTeamA and pointsTeamA

In [221]:
justFeatureDF = results[["rankTeamA", "pointsTeamA", 'teamA1', 'teamA2', 'teamA3', 'teamA4', 'teamA5', 'teamA6', 'teamA7', 'teamA8', 'teamA9', 'teamA10', 'teamA11', 'teamA12', 'teamA13', 'teamA14', 'teamB1', 'teamB2', 'teamB3', 'teamB4', 'teamB5', 'teamB6', 'teamB7', 'teamB8', 'teamB9', 'teamB10', 'teamB11', 'teamB12', 'teamB13', 'teamB14']]
features = results[["rankTeamA", "pointsTeamA", 'teamA1', 'teamA2', 'teamA3', 'teamA4', 'teamA5', 'teamA6', 'teamA7', 'teamA8', 'teamA9', 'teamA10', 'teamA11', 'teamA12', 'teamA13', 'teamA14', 'teamB1', 'teamB2', 'teamB3', 'teamB4', 'teamB5', 'teamB6', 'teamB7', 'teamB8', 'teamB9', 'teamB10', 'teamB11', 'teamB12', 'teamB13', 'teamB14']].values
print features[0]
clf = DecisionTreeClassifier(random_state=14)
scores = cross_val_score(clf, features, y_true,scoring=scorer)
print "Cross_Val_Score", np.mean(scores)
print features.shape
print features.dtype.names

[   1 4889    0    1    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    1    0    0    0    0    0    0    0]
Cross_Val_Score 0.564986309687
(1005L, 30L)
None


# Feature Importance

In [314]:
clf.fit(features, y_true)
clf.feature_importances_

array([ 0.04393278,  0.14730185,  0.03335284,  0.01512507,  0.12866377,
        0.1336197 ,  0.07727996,  0.09165517,  0.09630439,  0.        ,
        0.00776888,  0.00855614,  0.01069819,  0.00298227,  0.00434931,
        0.01515906,  0.00646508,  0.        ,  0.01803466,  0.01105764,
        0.00588201,  0.00851438,  0.0009608 ,  0.00525881,  0.00807905,
        0.00877404,  0.01346892,  0.01397509,  0.        ,  0.03417699,
        0.0039952 ,  0.0015393 ,  0.03507835,  0.0035474 ,  0.00180593,
        0.00263696,  0.        ])

## Rakesh, can you help me with this plot below. WHy is it not showing up? is it .plot(), .show()?

In [351]:
## Unable to show the plot here, any Idea Rakesh?
import matplotlib.pyplot as plt
import matplotlib
matplotlib.style.use('ggplot')
feature_importance = pd.Series(clf.feature_importances_, index=trainDfAfterEncodingAsFeatures.columns)
feature_importance.sort_values(inplace=True)
print feature_importance, trainDfAfterEncodingAsFeatures.columns
print len(trainDfAfterEncodingAsFeatures.columns)
feature_importance.plot(kind='barh', figsize=(7,6));
trainDfAfterEncodingAsFeatures.plot()
plt.show()


teamB14                   0.000000
teamA9                    0.000000
teamB6                    0.000000
teamA1                    0.000000
teamA14                   0.000961
teamB9                    0.001539
teamB12                   0.001806
teamB13                   0.002637
teamA5                    0.002982
teamB11                   0.003547
teamB8                    0.003995
teamA6                    0.004349
teamB1                    0.005259
teamA12                   0.005882
teamA8                    0.006465
teamA2                    0.007769
teamB2                    0.008079
teamA13                   0.008514
teamA3                    0.008556
teamB3                    0.008774
teamA4                    0.010698
teamA11                   0.011058
teamB4                    0.013469
teamB5                    0.013975
rankTeamB                 0.015125
teamA7                    0.015159
teamA10                   0.018035
RatingTeamA               0.033353
teamB7              

# Let's Plot the feature Importance Graph Before moving on to Random
# Forest Classifier

In [367]:
def graph_feature_importance(model, feature_names, autoscale=True, headroom=0.05, width=10, summarized_columns=None):
    if autoscale:
        x_scale = model.feature_importances_.max() + headroom
    else:
        x_scale =1
    
    feature_dict = dict(zip(feature_names, model.feature_importances_))
    
    if summarized_columns:
        for col_name in summarized_columns:
            sum_value = sum(x for i, x in feature_dict.iteritems() if col_name in i )
            keys_to_remove = [i for  i in feature_dict.keys() if col_name in i ]
            for i in keys_to_remove:
                feature_dict.pop(i)
            feature_dict[col_name] = sum_value
    
    resultsFig = pd.Series(feature_dict.values(), index=feature_dict.keys())
    resultsFig.sort(axis=1)
    resultsFig.plot(kind="barh", figsize=(width,len(resultsFig)/4), xlim=(0,x_scale))
#     fig = plt.figure()
#     ax = fig.add_subplot(111)
    
    plt.show()

categorical_variables = list(trainDfAfterEncodingAsFeatures.dtypes[trainDfAfterEncodingAsFeatures.dtypes == 'object'].index)
# trainDfAfterEncodingAsFeatures[categorical_variables].head()
graph_feature_importance(clf, trainDfAfterEncodingAsFeatures.columns, summarized_columns = categorical_variables)

C:\Anaconda2\lib\site-packages\ipykernel\__main__.py:18: FutureWarning: sort is deprecated, use sort_values(inplace=True) for for INPLACE sorting


# This above doesn't work since DecisionTreeClassifier' object has no attribute 'feature_importance_ and I should use this in RandomForest

## Trying RandomForest for experimenting

In [268]:
features

[0,
 0,
 0,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 ...]

In [373]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import roc_auc_score
model = RandomForestClassifier(n_estimators=50, oob_score=True, max_depth=3, max_leaf_nodes=5, min_samples_leaf=4)
model.fit(trainDf,y_true) # with only Ranks, Points, Rating, Homeground, Lastwins
# model.fit(trainDfAfterEncodingAsFeatures,y_true) #with only Ranks, Points, Rating, Homeground, Lastwins, AND Encoded features

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=3, max_features='auto', max_leaf_nodes=5,
            min_samples_leaf=4, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=50, n_jobs=1,
            oob_score=True, random_state=None, verbose=0, warm_start=False)

In [394]:
from sklearn import grid_search
parameters = {'n_estimators':(50, 100,500,1000),'max_depth':(3,5,7,10),'max_leaf_nodes':(2,4,5,7),'min_samples_leaf':(2,3,4,6)}
clf = grid_search.GridSearchCV(model, parameters)
clf.fit(trainDf,y_true)


GridSearchCV(cv=None, error_score='raise',
       estimator=RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=3, max_features='auto', max_leaf_nodes=5,
            min_samples_leaf=4, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=50, n_jobs=1,
            oob_score=True, random_state=None, verbose=0, warm_start=False),
       fit_params={}, iid=True, loss_func=None, n_jobs=1,
       param_grid={'n_estimators': (50, 100, 500, 1000), 'max_leaf_nodes': (2, 4, 5, 7), 'max_depth': (3, 5, 7, 10), 'min_samples_leaf': (2, 3, 4, 6)},
       pre_dispatch='2*n_jobs', refit=True, score_func=None, scoring=None,
       verbose=0)

In [334]:
clf.best_params_

AttributeError: 'DecisionTreeClassifier' object has no attribute 'best_params_'

In [374]:
print "oob_score", model.oob_score_
# print "oob_score", clf.oob_score_ #For grid_csv
print "score", model.score(trainDf,y_true) #Returns the coefficient of determination R^2 of the prediction.
# print "score", model.score(trainDfAfterEncodingAsFeatures,y_true) #Returns the coefficient of determination R^2 of the prediction.
cross_val = sklearn.cross_validation.cross_val_score(model, trainDf, y_true, cv=5)
# cross_val = sklearn.cross_validation.cross_val_score(model, trainDfAfterEncodingAsFeatures, y_true, cv=5)
# print cross_val_score.mean
# print np.mean(cross_val)
print np.mean(cross_val)


oob_score 0.644361833953
score 0.648079306072
0.623203742044


In [370]:
## Trying to plot the graph
# graph_feature_importance(model, trainDfAfterEncodingAsFeatures.columns, summarized_columns = categorical_variables)
graph_feature_importance(model, trainDf.columns, summarized_columns = categorical_variables)


807


C:\Anaconda2\lib\site-packages\ipykernel\__main__.py:18: FutureWarning: sort is deprecated, use sort_values(inplace=True) for for INPLACE sorting


## Let's predict the world cup matches


In [381]:
testDf.columns
# pool1=["New Zealand", "Australia", "Sri Lanka", "Bangladesh", "England", "Afghanistan", "Scotland"]
# pool2=["India", "South Africa", "Pakistan", "West Indies", "Ireland", "Zimbabwe", "United Arab Emirates"]

# import itertools
# list(itertools.combinations(pool1,2))

Index([u'rankTeamA', u'pointsTeamA', u'RatingTeamA', u'rankTeamB',
       u'pointsTeamB', u'RatingTeamB', u'home_groundOfAorB_1or2',
       u'teamALastWin', u'teamBLastWin'],
      dtype='object')

In [ ]:
teamRating = {"New Zealand":3, "Australia", "Sri Lanka", "Bangladesh", "England", "Afghanistan", "Scotland",
             "India", "South Africa", "Pakistan", "West Indies", "Ireland", "Zimbabwe", "United Arab Emirates"}


In [393]:
nz = [3,4284,116 ,5,5811,106, 1, True,False]
nzwi = []
SLSA = [5,5811,106, 4,5131,112,0, True,False]
InBa = [2,5278,117,7,3094,96,0,True,False]
SriNz = [5,5811,106, 3,4284,116,1,False,True]
NzAus = [3,4284,116,1,4889,129,2,True,True]
model.predict(NzAus)
# a =  [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
# b =  [0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0]
# input_for_predicting = [3, 4, a[0], a[1], a[2],a[3],a[4],a[5],a[6],a[7],a[8],a[9],a[10],a[11],a[12],a[13], b[0],b[1],b[2],b[3],b[4],b[5],b[6],b[7],b[8],b[9],b[10],b[11],b[12],b[13]]
# clf.predict(input_for_predicting)

array([1], dtype=int64)

## Rakesh, need help in calculating this C-Stat Score.
- U must be wondering why am I interested in C-Stat score with RandomForest when I already have cross_val_score from DecisionTree.
- This is because, I have the intuition that RandomForest might give me a better score and I will have better score to show to leman. If I go ahead with just the cross_val_score, I will have to lie as the score from Decision tree is as low as 0.57

In [185]:
y_oob = model.oob_prediction_
# print "C-Stat: ", roc_auc_score(y_true,y_oob)   #Getting error with, some issue in type of arguments passed. Rakesh, Could u check once?
print y_oob

[ 0.92986263  0.98048473  0.91310385 ...,  1.05816704  1.38981526
  1.6015873 ]


array([ 0.14060507,  0.20489716,  0.00219963,  0.00997314,  0.01728387,
        0.0208175 ,  0.01019713,  0.01727822,  0.00676036,  0.00436816,
        0.00067557,  0.01132397,  0.01725618,  0.00514591,  0.01706523,
        0.00916741,  0.01341935,  0.03935191,  0.02933408,  0.0265886 ,
        0.0393802 ,  0.01896208,  0.05096352,  0.03031806,  0.02551364,
        0.05864157,  0.04019553,  0.015927  ,  0.03589662,  0.08049332])

## Could Draw a graph out of this

In [181]:
def describe_categorical(results):
    from IPython.display import display, HTML
    display(HTML(results[results.columns[results.dtypes == "object"]].describe().to_html()))

describe_categorical(results)

## Before I predict, I should know the featue values.
- WIll calculate the encoded values of each team and keep them as list in a dictionary
- will create a dict of teamA rankings and teamA points

In [109]:
a =  [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
b =  [0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0]
input_for_predicting = [3, 4, a[0], a[1], a[2],a[3],a[4],a[5],a[6],a[7],a[8],a[9],a[10],a[11],a[12],a[13], b[0],b[1],b[2],b[3],b[4],b[5],b[6],b[7],b[8],b[9],b[10],b[11],b[12],b[13]]
clf.predict(input_for_predicting)

array([2], dtype=int64)

## So good news, the predictor is working fine.
- I created a dict for all teams to hold the encoded value

In [145]:
# distinct_teams = results["teamA"].unique()

dictTeamsEncoding={}

for i in range(len(results["teamA"])):
    team = results["teamA"][i]
    if team not in dictTeamsEncoding:
        # find the encoding
        encodedTeam = results.ix[i,16:30].tolist()
        dictTeamsEncoding[team] = encodedTeam
    if len(dictTeamsEncoding) == 14:
        break

print dictTeamsEncoding

{'Pakistan': [0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0], 'Australia': [0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], 'England': [0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], 'Bangladesh': [0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], 'South Africa': [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0], 'Afghanistan': [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], 'Scotland': [0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0], 'India': [0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0], 'Sri Lanka': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0], 'Zimbabwe': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1], 'Ireland': [0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0], 'United Arab Emirates': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0], 'West Indies': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0], 'New Zealand': [0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0]}


## Now going to store the values of team's Rank and Points
- Great, the dicts have been initialized

In [163]:
dictRankTeam = {}
dictPointsTeam = {}

for i in range(len(results["teamA"])):
    team = results["teamA"][i]
    if team not in dictRankTeam:
        # find both the points and rank and put them in respective dictionaries 
        rank = int(results.ix[i,4].tolist())
        dictRankTeam[team] = rank
        
        points = int(results.ix[i,5].tolist())
        dictPointsTeam[team] = points
        
    if len(dictPointsTeam) == 14:
        break

print dictRankTeam
print "\n"
print dictPointsTeam

{'Pakistan': 9, 'Australia': 1, 'England': 6, 'Bangladesh': 7, 'South Africa': 4, 'Afghanistan': 12, 'Scotland': 14, 'India': 2, 'Sri Lanka': 5, 'Zimbabwe': 11, 'Ireland': 10, 'United Arab Emirates': 13, 'West Indies': 8, 'New Zealand': 3}


{'Pakistan': 3721, 'Australia': 4889, 'England': 3968, 'Bangladesh': 3094, 'South Africa': 5131, 'Afghanistan': 618, 'Scotland': 200, 'India': 5278, 'Sri Lanka': 5811, 'Zimbabwe': 1250, 'Ireland': 549, 'United Arab Emirates': 300, 'West Indies': 2471, 'New Zealand': 4284}


## Now we will take all 14 teams and put the teams in their pools and then apply 7c2 as each team will play with every other team in that pool and the best 4 will advance.

In [237]:
pool1=["New Zealand", "Australia", "Sri Lanka", "Bangladesh", "England", "Afghanistan", "Scotland"]
pool2=["India", "South Africa", "Pakistan", "West Indies", "Ireland", "Zimbabwe", "United Arab Emirates"]

import itertools
list(itertools.combinations(pool1,2))



[('New Zealand', 'Australia'),
 ('New Zealand', 'Sri Lanka'),
 ('New Zealand', 'Bangladesh'),
 ('New Zealand', 'England'),
 ('New Zealand', 'Afghanistan'),
 ('New Zealand', 'Scotland'),
 ('Australia', 'Sri Lanka'),
 ('Australia', 'Bangladesh'),
 ('Australia', 'England'),
 ('Australia', 'Afghanistan'),
 ('Australia', 'Scotland'),
 ('Sri Lanka', 'Bangladesh'),
 ('Sri Lanka', 'England'),
 ('Sri Lanka', 'Afghanistan'),
 ('Sri Lanka', 'Scotland'),
 ('Bangladesh', 'England'),
 ('Bangladesh', 'Afghanistan'),
 ('Bangladesh', 'Scotland'),
 ('England', 'Afghanistan'),
 ('England', 'Scotland'),
 ('Afghanistan', 'Scotland')]

## Let's add the team names as the features by 1-of-K encoding or 1-hot encoding
- We are going to add the column of "TeamARanksHigher" / "TeamBRanksHigher" for both teams A and B
- Though we have a column already present as rankTeamA, and rankTeamB, but let's add the above two and see if there are notable changes by converting them to boolean.

## Use OneHot encoder

## Let's evaluate the performance after one hot encoding and the cross_val_score using only the team names as features

## Starting here, we are taking a new approach
### We will just keep adding the new features including this hot key encoding to results dataframe.
- Then we will use combination to judge which is giving the best possible model
- Or you know what let's keep saving the score in a list as tuples (score, no. of features considered)

In [ ]:
## Dropped this Idea

## New approach ends here

## Let's Select the best model.
### The best Model would be one which will selectively choose those features with inclusion of which the cross_val_score will be increased.
- My approach here is to take a initial baseline model, keep looking for the best available feature to be added next from the remaining feature list.
- keep doing this untill either all the features have been included or the score is not improving when going from one iteration to other.

## Let's actually begin Model selection with available feature list

### Let's programatically fill the dataframe appropriately the encoding

### Yes, the baseline model is now ready with the initial two features as our encoded team names

# Score with Base Line Model (encoded team names as feature) = 0.5491